accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,496 @@
+
+The rationale for making apphost EXEs opt-in is that they are a non-portable asset, creating a strange asymmetry in an otherwise portable app. Perhaps, we should make it easy for developers to generate an apphost only for debug builds to aid the development experience. That said, if developers want a better development experience, they could just as easily build RID-specific apps (the proposed default) for development and then build a portable app with no EXE for production where an EXE is presumably not needed.
+
+Note: Portable apps are not a meaningful option for client apps since almost all client apps require an executable launcher, and there is only ever one generated. Portable apps are only meaningful for commandline apps.
+
+Note: In theory, we could consider an option to create multiple apphosts for different RIDs, either of different names or in different directories. Again, this capability would need to be opt-in.
+
+### Rationalize RID folders for RID-specific and portable apps
+
+There have been some issues filed (see appendix) that suggest that managed libraries are not resilient to RID-specific libraries changing location as a result of changing build type (portable to RID-specific). This would also be true for single file deployment builds.
+
+We could enable (as an option) creating `runtimes` folders for RID-specific apps to make the folder layout the same across deployment types. Doing that would put single file at a significant disadvantage. We'd rather encourage the ecosystem to support both folder layouts. If we had to choose one of these deployments types to disadvantage, it would be portable (with the `runtimes` folders), not RID-specific, which is the likely the opposite of what developers are tuning for today.
+
+We need to invest more in docs and tools to the end of clear and actionable guidance.
+
+### Use SDK RID as default RID
+
+RID-specific apps should default to the SDK RID as an implicit RID, and no longer require a RID to be specified. The SDK already does this for the apphost for portable apps. This proposal is a straightforward extension of that behavior.
+
+This means that the following commands would produce assets in terms of the implicit RID, such as with the following commands.
+
+- `dotnet publish --self-contained`
+- `dotnet publish /p:PublishSingleFile=true`
+
+Note: As already suggested, this change would make Dockerfiles MUCH easier to write. Needing to specify RIDs is probably the worst part of using .NET with Docker.
+
+Specifying a RID should have no effect on the build type/deployment model. It should simply refine the existing model. This change largely follows from using the SDK RID as the implicit default for all types of RID-specific apps. When a RID is specified, it is substituted for the implicit RID.
+
+This means that the following commands would produce a RID-specific framework-dependent app:
+
+- `dotnet build -r linux-x64`
```

Comment:
How much thought has been put forward on `linux-x64`? Unlike Windows, there can be a lot of differences between the various distros and writing a portable Linux app can be difficult.

In some cases, you might have different native dependencies for Ubuntu vs RHEL vs SLES vs etc (and sometimes even different dependencies between each of these due to differences in what ships in the system by default).

Reply:
Very little thought. I'd heard very few requests for build-time differentiation. We are forced to do it for Alpine (musl libc), but that's logically separate from what you are getting at. Certainly, we could extend the RID graph for distros. We need compelling scenarios.

Reply:
One scenario that comes to my mind is dealing with openssl since there can be different incompatible versions/variants between distros and versions of the same distro.

Reply:
We already have that problem solved. Put another way, I don't think enable `linux` as any any-style RID would change our approach to OpenSSL. Same with ICU. Perhaps I'm missing something.

Reply:
We have solved it for the .NET Runtime dependencies on icu and openssl by using non-standard mechanisms to bind to these libraries. The solution was non-trivial to setup and requires constant maintenance. (In theory, we could have used similar non-standard mechanism to bridge the glibc/musl differences too, but that was too complex even for our taste, so we choose to let glibc/musl bleed out in the UX.)

We have not solved it for 3rd party packages with native libraries that depend on openssl, icu and other native libraries. For example, the dependency on libtinfo was giving me a hard time in one of my projects. I think it is rare for authors of 3rd party packages with native dependencies to go through the pains of targeting generic linux RID. The authors of 3rd party packages with native dependencies typically end up picking up a set of Linux distros, configurations and architectures to build and package the native dependencies for. If your project target platform is lucky to be on this list, great! If it is not, you are on a unhappy path.

The only scalable way to solve this is to add fallback to build-from-source for the native dependencies (https://github.com/dotnet/designs/pull/233#issuecomment-877676493). Fallback to build-from-source would work well for current-RID-specific publish, so this proposal is good for it. It would not work well (without additional manual configuration) for RID-neutral publish.